src/transport/transport.cc

```diff
@@ -466,87 +480,156 @@ void ConvectionTransport::zero_time_step()
+    START_TIMER("time step rescaling");
+    bool scaled = false; //flag to avoid rescaling newly created bc term
+    // if FLOW or DATA or BC changed ---------------------> recompute boundary condition
+    if ( (mh_dh->time_changed() > transport_bc_time)
+        || data_.porosity.changed()
+        || data_.bc_conc.changed() )
+    {
         set_boundary_conditions();
-        // scale boundary sources
-        for (unsigned int sbi=0; sbi<n_subst_; sbi++) VecScale(bcvcorr[sbi], time_->estimate_dt());
-
-        need_time_rescaling = true;
-    } else {
-        // possibly read boundary conditions
-        if (data_.bc_conc.changed() ) {
-            set_boundary_conditions();
-            // scale boundary sources
-            for (unsigned int  sbi=0; sbi<n_subst_; sbi++) VecScale(bcvcorr[sbi], time_->dt());
-        }
+        // rescale by time step
+        DBGMSG("BC - rescale NEW dt.\n");
+        for (unsigned int  sbi=0; sbi<n_subst_; sbi++) 
+            VecScale(bcvcorr[sbi], dt_new);
+        scaled = true;
     }
+    
+    if( !scaled && time_->is_changed_dt() ) // if time step changed, only rescale
+    {
+        DBGMSG("BC - rescale SCALE dt.\n");
+        for (unsigned int  sbi=0; sbi<n_subst_; sbi++) 
+            VecScale(bcvcorr[sbi], dt_scaled);
+    }
+    
 
-    if (need_time_rescaling) {
-        if ( is_convection_matrix_scaled ) {
-            // rescale matrix
-            MatShift(tm, -1.0);
-            MatScale(tm, time_->estimate_dt()/time_->dt() );
-            MatShift(tm, 1.0);
-
-            for (unsigned int sbi=0; sbi<n_subst_; sbi++) VecScale(bcvcorr[sbi], time_->estimate_dt()/time_->dt());
-
-        } else {
-            // scale fresh convection term matrix
-            MatScale(tm, time_->estimate_dt());
-            MatShift(tm, 1.0);
-            is_convection_matrix_scaled = true;
-
+    // if DATA or TIME STEP changed -----------------------> rescale source term
+    scaled = false; //reset; flag to avoid rescaling newly created src term
+    if( !is_src_term_scaled ) { //if it is not scaled (freshly computed)
+        DBGMSG("SRC - rescale NEW dt.\n");
+        for (unsigned int sbi=0; sbi<n_subst_; sbi++)
+        {
+            VecScale(v_sources_corr[sbi], dt_new);
+            VecScale(v_tm_diag[sbi], dt_new);
         }
-        need_time_rescaling = false;
+        is_src_term_scaled = true;
+        scaled = true;
+    }
+    if( !scaled && time_->is_changed_dt() ) { // if time step changed, only rescale
+        DBGMSG("SRC - rescale SCALE dt.\n");
+        for (unsigned int sbi=0; sbi<n_subst_; sbi++)
+        {
+            VecScale(v_sources_corr[sbi], dt_scaled);
+            VecScale(v_tm_diag[sbi], dt_scaled);
+        }
+    }
+    
+    // if DATA or TIME STEP changed -----------------------> rescale transport matrix
+    scaled = false; //reset; flag to avoid rescaling newly created tm term
+    if ( !is_convection_matrix_scaled ) { // scale fresh convection term matrix
```

Comment:
Is there still reason to not perform scaling at matrix assembly? Since the reason was that we did not know time step at assembly, but now we should know it. Same for BC and sources above.